poppy/sub_sampled_optics.py

```diff
@@ -0,0 +1,253 @@
+import numpy as np
+import poppy
+import matplotlib.pyplot as plt
+import astropy.units as u
+
+from .poppy_core import OpticalElement, Detector, Wavefront, PlaneType, _PUPIL, _IMAGE, _RADIANStoARCSEC
+from .optics import CircularAperture
+from .utils import measure_centroid
+
+import logging
+_log = logging.getLogger('poppy')
+
+class subapertures(poppy.OpticalElement):
+        """
+        Example roadmap:
+        
+        #generate wavefront
+        wf = poppy.wavefront() #or fresnel wavefront
+        
+        #...various surfaces
+        
+        subaperture(ARRAY OF OPTICS) 
+        #initialize this new class, where the array of optics define the subapertures (e.g. lenslets)
+        
+        subapertures.sample_wf(wf) #this function takes the wavefront and subsamples it by the area of each optic
+        subapertures.get_wavefront_array() #returns an array of input sub-wavefronts multipled by subaperture optics
+        subapertures.get_psfs() #fraunhofer or fresnel propagation of each pupil  to the image/ waist
+        image=subapertures.get_composite_wavefont() # returns wavefront of image plane of all the spots put back together
+        subapertures.opd #single array made up of subapertures
+        subapertures.amplitude  #single array made up of subapertures
+        subapertures.getphasor #returns propagated array of spots from get_composite_wavefont
+        
+        
+        Parameters
+        ----------
+        
+        crosstalk: boolean
+                  this variable sets whether light can leak from one lenslet image plane into the neighbors.
+                  Default is False.
+        x_y_offset: tuple
+                  offset of central grid vertex from center of incident wavefront
+        input_wavefront: None or poppy.Wavefront 
+                  : the unsampled wavefront incident on the subapertures
+        """
+        
+        def __init__(self,
+                     dimensions = (2,2),
+                     optic_array = np.array([[CircularAperture(radius=2.,planetype=PlaneType.pupil),
```

Comment:
Watch out for mutable default arguments! (see http://stackoverflow.com/questions/1132941/least-astonishment-in-python-the-mutable-default-argument for more details)